src/Symfony/Component/HttpClient/Internal/AmpClientState.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient\Internal;
+
+use Amp\Http\Client\Connection\DefaultConnectionFactory;
+use Amp\Http\Client\Connection\LimitedConnectionPool;
+use Amp\Http\Client\Connection\UnlimitedConnectionPool;
+use Amp\Http\Client\HttpClient;
+use Amp\Http\Client\HttpClientBuilder;
+use Amp\Socket\Certificate;
+use Amp\Socket\ClientTlsContext;
+use Amp\Socket\ConnectContext;
+use Amp\Socket\DnsConnector;
+use Amp\Socket\StaticConnector;
+use Amp\Sync\LocalKeyedSemaphore;
+
+/**
+ * Internal representation of the Amp client's state.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @internal
+ */
+final class AmpClientState extends ClientState
+{
+    /** @var HttpClientBuilder[] */
+    private $clients = [];
+    private $maxHostConnections = 0;
+    private $builder;
+
+    public function __construct(?HttpClientBuilder $builder, int $maxHostConnections)
+    {
+        $this->builder = ($builder ?? (new HttpClientBuilder())->allowDeprecatedUriUserInfo())->followRedirects(0);
+        $this->maxHostConnections = $maxHostConnections;
+    }
+
+    public function getClient(array $options): HttpClient
+    {
+        $options = [
+            'bindto' => $options['bindto'] ?: '0',
+            'verify_peer' => $options['verify_peer'],
+            'capath' => $options['capath'],
+            'cafile' => $options['cafile'],
+            'local_cert' => $options['local_cert'],
+            'local_pk' => $options['local_pk'],
+            'ciphers' => $options['ciphers'],
+            'capture_peer_cert_chain' => $options['capture_peer_cert_chain'] || $options['peer_fingerprint'],
+        ];
+
+        $key = implode("\0", $options);
+
+        if (isset($this->clients[$key])) {
+            return $this->clients[$key];
+        }
+
+        $context = new ClientTlsContext('');
+        $options['verify_peer'] || $context = $context->withoutPeerVerification();
+        $options['cafile'] && $context = $context->withCaFile($options['cafile']);
+        $options['capath'] && $context = $context->withCaPath($options['capath']);
+        $options['local_cert'] && $context = $context->withCertificate(new Certificate($options['local_cert'], $options['local_pk']));
+        $options['ciphers'] && $context = $context->withCiphers($options['ciphers']);
+        $options['capture_peer_cert_chain'] && $context = $context->withPeerCapturing();
+
+        if ($options['bindto']) {
+            $connector = (file_exists($options['bindto']) ? 'unix://' : 'tcp://').$options['bindto'];
+            $connector = new StaticConnector($connector, new DnsConnector());
```

Comment:
You could use `Amp\Socket\connector()` here instead of `new DnsConnector()` to use the default configured global connector. I originally wanted to suggest this to improve the DNS cache, but the DNS cache is separately anyway.